sycl/doc/extensions/DataFlowPipes/data_flow_pipes.asciidoc

```diff
@@ -0,0 +1,650 @@
+
+If _host_writeable_ is true, then the host program writes to the pipe and the kernel also accessing the pipe must read from it.  If _host_writeable_ is false, then the host program reads from the pipe, and the kernel accessing the pipe must write to it.
+
+Mapping a pipe does not impact the _min_capacity_ property of the pipe in any way, so a mapped memory region used to read from or write to the pipe from the host must not be considered as adding additional capacity to the pipe from the perspective of preventing application deadlock.
+
+The APIs are defined as:
+
+[cols="2*^",options="header",stripes=none]
+|===
+|Function |Description  
+|`template <pipe_property::writeable host_writeable> +
+dataT* map(size_t requested_size, size_t &mapped_size);`
+|Returns a _dataT *_ in the host address space. The host can write data to this pointer for reading by a device pipe endpoint, if it was created with template parameter `host_writeable = true`.  Alternatively, the host can read data from this pointer if it was created with template parameter `host_writeable = false`.
+
+The value returned in the mapped_size argument specifies the maximum number of bytes that the host can access. The value specified by _mapped_size_ is less than or equal to the value of the _requested_size_ argument that the caller specifies.  _mapped_size_ does not impact the _min_capacity_ property of the pipe.
+
+After writing to or reading from the returned _dataT *_, the host must execute one or more `unmap` calls on the same pipe, to signal to the runtime that data is ready for transfer to the device (on a write), and that the runtime can reclaim the memory for reuse (on a read or write). If `map` is called on a pipe before `unmap` has been used to unmap all memory mapped by a previous `map` operation, the buffer returned by the second `map` call will not overlap with that returned by the first call.
+
+|`static size_t unmap(T* mapped_ptr, size_t size_to_unmap);`
+|Signals to the runtime that the host is no longer using _size_to_unmap_ bytes of the host allocation that was returned previously from a call to `map`. In the case of a writeable host pipe, calling `unmap` allows the unmapped data to become available to the kernel. If the _size_to_unmap_ value is smaller than the _mapped_size_ value specified to `map`, then multiple `unmap` function calls are necessary to unmap the full capacity of the host allocation. It is legal to perform multiple `unmap` function calls to unmap successive bytes in the buffer returned by `map`, up to the _mapped_size_ value defined in the `map` operation.
+|===
```

Comment:
These functions are useful but I find the interface awkward. Perhaps because it does not feel like modern C++? :-)

It looks very difficult to provide a zero-copy behavior if the OS allows it and you have already the data somewhere.
Perhaps an API with `read` and `write` taking a `std::range` and returning the number of elements successfully read or written?

Reply:
To give an idea of what features could be exposed, my colleague @sonals mentioned our low-level API https://xilinx.github.io/XRT/master/html/streaming_data_support.html

The polling time could be provided in the API as an optional parameter.

Reply:
Tracked for future in https://github.com/intel/llvm/issues/832

Reply:
Tracked for future in https://github.com/intel/llvm/issues/832